feedhenry-templates/sync-ios-swift
# RHMAP-21014 - Add error handling for updates

## Motivation

Add error handling on client.
Handle the cases when error happened on save and when conflict happened.

## Notes

Clients should additionally setup hashing method on the server that ignores fields like updatedAt etc. This fields usually clash with the hashing mechanism causing conflicts when subsequent updates for same id are applied at the same time.
We cannot do that on framework level as this depends on the actual application data.

## Testing

### Pre-requirements:

#### The test needs to be done with this template locally. For it clone the project and check this PR as follows. 

```
cmacedo@camilas-MBP ~/redhat/sync-ios-swift (master) $ git remote add upstream git@github.com:feedhenry-templates/sync-ios-swift.git
cmacedo@camilas-MBP ~/redhat/sync-ios-swift (master) $ git fetch upstream pull/37/head:pr
From github.com:feedhenry-templates/sync-ios-swift
 * [new ref]         refs/pull/37/head -> pr
cmacedo@camilas-MBP ~/redhat/sync-ios-swift (master) $ git checkout pr
Switched to branch 'pr'
cmacedo@camilas-MBP ~/redhat/sync-ios-swift (pr) $ 
``` 

#### Install and build the client application locally ( For it is required to be in the iOS env with Xcode )

```
$sudo gem install cocoapods
$pod install
```

#### Following the steps of the test. 

1) Go to the file `sync-ios-swift/sync-ios-app/fhconfig.plist` or via scode and change the configuration following an example. ( Use the data in the JIRA RHMAP-21014 )

```
<plist>
    <dict>
        <key>host</key>
        <string>hostTestURL</string>
        <key>appid</key>
        <string>kn2r4a4j4noaygmlq6fafda5</string>
        <key>projectid</key>
        <string>kn2r4a4vwznwaonfop4k4rez</string>
        <key>appkey</key>
        <string>9864d01845ba3d83cb1a9675ef22e5aee14b256d</string>
        <key>connectiontag</key>
        <string>0.0.2</string>
    </dict>
</plist>
```

*Note*: To emulate it in the Xcode use `sync-ios-app.xcworkspace`

2. Create "new" record
3. Update record to "new updated" in client
4. Wait for error in cloud app log
5. Observe that client receives error that data failed to update.
6. Observe that client data is being reverted.
7. Attempt to save record again as "test3"
8. Wait for error in cloud app log
9. Observe that client receives error that data failed to update in the logs. Following an example. 

```
{"_id":"5b60c1baa0c8ab5db58ddfc3","action":"update","cuid":"5A32C59C-2E52-4AF4-A972-0922C5E4A2F5","hash":"e73cfc2fee3f7b3fba6f5be3e21112768ece900a","uid":"5b60c1b2cb1f20dd5703fbb9","msg":"'error'","type":"failed","timestamp":1533067706949}
onSyncMessage::REMOTE_UPDATE_FAILED_MESSAGE
::onDataUpdated::refresh tableview
```

**Note:** Sync on this server is really slow. It may take serious amount of time to receive updates/failures. You can change `conf?.syncFrequency = 30` to `conf?.syncFrequency = 5` to speed things up.



## Notes

@shannon As issue was happening because of our template did not had functions for handling error cases implemented. Some flags were also missing.
There are no changes in actual SDK so users **will need to replicate this changes** to their application manually.
